### PR TITLE
Generate thumbnails on imported scenes.

### LINKED ISSUE #8760: Display scene resource thumbnails in the FileSystem dock
### Describe the project you are working on

3D Game with multiple levels using pre-made game assets from Synty and other asset packs.

### Describe the problem or limitation you are having in your project

To create levels quickly and reduce time, I really need to see the 3D models I have imported to be displayed as thumbnails in the FileSystem window. This would allow me to easily find the correct asset without having to double-click the 3D model to open up the 'Advanced Import Settings' to see which model is which.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

Having the ability to 'see' which model is which, instead of having to read carefully each file name of each 3D model or having to open the 'Advanced Import Settings' window would save a lot of time creating levels. GridMaps aren't a good solution as the models have to fit into a specific grid. Most 3D levels have arbitrary sized models, unlike a 2D game where image assets are typicaly 16x16, 32x32 etc.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

In the FileSystem view...
Screenshot 1

or in it's own Horizontal tab...
Screenshot 2

### If this enhancement will not be used often, can it be worked around with a few lines of script?

I'm not sure how easy it would be to script, however we can already display thumbnails in GridView, so the code should be there I'd imagine.

### Is there a reason why this should be core and not an add-on in the asset library?

this should be core functionality. If it was a plugin it could easily be broken in updates.

### PR DESCRIPTION
**Pull request by SaracenOne**
godot windows editor dev x86_64_OmkGxF9AY3

Ends a somewhat long-standing discontinuity of the imported scenes (FBX, GLTF, Collada, ect.) not having any thumbnail previews like regular Godot scenes. It does this by providing a hook step where they can be generated and cached upon reimport.

Keeping this in draft for now since it could probably benefit from further refactoring, improving bounds calculations for skinned meshes, and adding displayable gizmos for the cases where no mesh instances could be found (animation or collision-only imports, for example).

- *Production edit: This closes https://github.com/godotengine/godot-proposals/issues/8760.*

### CODE REVIEW

**Comment by JekSun97** [comment]
In editor_inteface.cpp error in the comment on line 300

File: editor/editor_interface.cpp [hunk 1]
```diff
@@ -195,6 +221,132 @@ Vector<Ref<Texture2D>> EditorInterface::make_mesh_previews(const Vector<Ref<Mesh
...
+
+	light->set_transform(xform * Transform3D().looking_at(Vector3(-2, -1, -1), Vector3(0, 1, 0)));
+	light2->set_transform(xform * Transform3D().looking_at(Vector3(+1, -1, -2), Vector3(0, 1, 0)));
+
+	// Update the renderer to get the screenshot.
+	DisplayServer::get_singleton()->process_events();
+	Main::iteration();
+	Main::iteration();
+
+	// Get the exture.
```
Comment (SpazElectro) [inline editor/editor_interface.cpp:300]: texture*

**Comment by fire** [comment]
@SaracenOne Do you think this can be done for the 4.4 cycle?

**Review by Calinou — COMMENTED** [review]
Tested locally (rebased on top of `master` https://github.com/godotengine/godot/commit/f952bfe9985ad8f507cc29b2c7601bbba18b8039), it works as expected.

1 | 2 | 3 | 4
-|-|-|-
image | image | image | image

Note that you must reimport the scene for the preview to be generated (existing projects won't have one automatically created). Also, it can take a few seconds for the icon to appear in the FileSystem dock after reimporting.

For some assets, it results in a gray preview, presumably due to skinning resulting in a different AABB than needed. This happens with `player.glb` from the [3D Platformer demo](https://github.com/godotengine/godot-demo-projects/tree/master/3d/platformer):

image

Also, `cube.glb` from the [Global Illumination demo](https://github.com/godotengine/godot-demo-projects/tree/master/3d/global_illumination) is too zoomed in:

image

`zdm2.glb` from the same demo is fine though:

image

**Comment by fire** [comment]
Going to do a rebase and push it.

**Comment by fire** [comment]
The code is currently failing at:

```
 ===============================================================================
./modules/gltf/tests/test_gltf_extras.h:104:
TEST CASE:  [SceneTree][Node] GLTF test mesh and material meta export and import

./modules/gltf/tests/test_gltf_extras.h:104: FATAL ERROR: test case CRASHED: SIGSEGV - Segmentation violation signal

===============================================================================
[doctest] test cases:     7 |     6 passed | 1 failed | 1076 skipped
[doctest] assertions: 56742 | 56742 passed | 0 failed |
[doctest] Status: FAILURE!
[1] ./bin/godot.linuxbsd.editor.dev.x86_64.llvm.san(+0x20a5f885) [0x55c219d2f885] (/home/runner/work/godot/godot/platform/linuxbsd/crash_handler_linuxbsd.cpp:61)
[2] /lib/x86_64-linux-gnu/libc.so.6(+0x42520) [0x7f5f0a642520] (??:0)
[3] Node::add_child(Node*, bool, Node::InternalMode) (/home/runner/work/godot/godot/scene/main/node.cpp:1644)
[4] EditorInterface::make_scene_preview(String const&, Node*, int) (/home/runner/work/godot/godot/editor/editor_interface.cpp:246)
[5] ResourceImporterScene::_generate_preview_for_scene(String const&, Node*) (/home/runner/work/godot/godot/editor/import/3d/resource_importer_scene.cpp:2820)
[6] ResourceImporterScene::import(long, String const&, String const&, HashMap<StringName, Variant, HashMapHasherDefault, HashMapComparatorDefault<StringName>, DefaultTypedAllocator<HashMapElement<StringName, Variant> > > const&, List<String, DefaultAllocator>*, List<String, DefaultAllocator>*, Variant*) (/home/runner/work/godot/godot/editor/import/3d/resource_importer_scene.cpp:3174)
[7] ./bin/godot.linuxbsd.editor.dev.x86_64.llvm.san(+0x22a36a7d) [0x55c21bd06a7d] (/home/runner/work/godot/godot/./modules/gltf/tests/test_gltf_extras.h:94)
[8] ./bin/godot.linuxbsd.editor.dev.x86_64.llvm.san(+0x22699e96) [0x55c21b969e96] (/home/runner/work/godot/godot/./modules/gltf/tests/test_gltf_extras.h:138)
[9] doctest::Context::run() (/home/runner/work/godot/godot/./thirdparty/doctest/doctest.h:7008)
[10] test_main(int, char**) (/home/runner/work/godot/godot/tests/test_main.cpp:258)
[11] Main::test_entrypoint(int, char**, bool&) (/home/runner/work/godot/godot/main/main.cpp:898)
[12] ./bin/godot.linuxbsd.editor.dev.x86_64.llvm.san(main+0x268) [0x55c219d2e958] (/home/runner/work/godot/godot/platform/linuxbsd/godot_linuxbsd.cpp:68)
[13] /lib/x86_64-linux-gnu/libc.so.6(+0x29d90) [0x7f5f0a629d90] (??:0)
[14] /lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0x80) [0x7f5f0a629e40] (??:0)
[15] ./bin/godot.linuxbsd.editor.dev.x86_64.llvm.san(_start+0x25) [0x55c219c6e465] (??:?)
-- END OF BACKTRACE --
================================================================
/home/runner/work/_temp/7e6d1766-f780-4f71-9362-f34b172819ee.sh: line 3:  2616 Aborted                 (core dumped) ./bin/godot.linuxbsd.editor.dev.x86_64.llvm.san --headless --test --force-colors```

**Comment by fire** [comment]
We discussed how the thumbnail in imported scenes needs features added.

1. AABB estimation related to skinned meshes has also proven to be a pain point for the advanced importer. 
2. Provide an API that could allow more thumbnail generation customization for regular Godot scenes.

What do you think about greenlighting the pull request as is and revisiting thumbnails in imported scenes feature enhancements later to see if it can be improved?

**Review by fire — APPROVED** [review]
Please take a look at the previous comment for the summary.

@Calinou mentioned that we should not wait for the entire feature list to be implemented since it's already an improvement over what we have now.

**Review by AThousandShips — APPROVED** [review]

**Review by Repiteo — APPROVED** [review]
Accepting with the expectation that this will be followed up on soon™ with the aforementioned features

**Comment by Repiteo** [comment]
Thanks!

**Pull request merged**